File: HangTen/Models/MotherboardWorkoutRecorder.swift
import Foundation

struct MotherboardWorkoutRecorder {
    private struct StepState {
        let plannedActiveDuration: TimeInterval
        let stepStartElapsed: TimeInterval
        var intervals: [LoadInterval] = []
        var openStart: TimeInterval?
        var pendingStart: TimeInterval?
        var pendingRelease: TimeInterval?
        var peakLoadKGF: Double?
        var sampleCount = 0
        var status: WorkoutStepMeasurement.Status = .unmeasured
        var permitsMergeWithPrevious = true

        func clippedTime(_ time: TimeInterval) -> TimeInterval {
            min(time, plannedEnd)
        }

        var plannedEnd: TimeInterval {
            stepStartElapsed + plannedActiveDuration
        }
    }

    private let configuration: MotherboardDetectionConfiguration
    private var states: [String: StepState] = [:]
    private var stepOrder: [String] = []
    private(set) var currentStepID: String?

    private(set) var currentLoadKGF: Double?
    private(set) var currentPeakLoadKGF: Double?
    private(set) var currentLoadedDuration: TimeInterval = 0

    var isLoaded: Bool {
        guard let currentStepID else { return false }
        return states[currentStepID]?.openStart != nil
    }

    init(configuration: MotherboardDetectionConfiguration = .init()) {
        self.configuration = configuration
    }

    mutating func consume(
        _ measurement: MotherboardMeasurement,
        stepID: String,
        plannedActiveDuration: TimeInterval,
        workoutElapsed: TimeInterval,
        stepStartElapsed: TimeInterval? = nil,
        isActive: Bool
    ) {
        currentLoadKGF = measurement.aggregateLoadKGF
        if currentStepID != stepID {
            currentPeakLoadKGF = nil
            currentLoadedDuration = 0
        }
        currentStepID = stepID

        var state = state(
            for: stepID,
            plannedActiveDuration: plannedActiveDuration,
            stepStartElapsed: stepStartElapsed ?? workoutElapsed
        )
        guard isActive else {
            states[stepID] = state
            return
        }

        state.sampleCount += 1
        state.peakLoadKGF = max(state.peakLoadKGF ?? measurement.aggregateLoadKGF, measurement.aggregateLoadKGF)
        currentPeakLoadKGF = state.peakLoadKGF

        let time = state.clippedTime(workoutElapsed)
        let isQualifying = measurement.aggregateLoadKGF >= configuration.thresholdKGF
        let isReleased = measurement.aggregateLoadKGF < configuration.thresholdKGF * configuration.releaseRatio

        if state.openStart == nil {
            if isQualifying {
                if let pendingStart = state.pendingStart {
                    if time - pendingStart >= configuration.debounceDuration {
                        openInterval(in: &state, at: pendingStart)
                    }
                } else {
                    state.pendingStart = time
                    if configuration.debounceDuration == 0 {
                        openInterval(in: &state, at: time)
                    }
                }
            } else {
                state.pendingStart = nil
            }
        } else if isReleased {
            if let pendingRelease = state.pendingRelease {
                if time - pendingRelease >= configuration.debounceDuration {
                    closeInterval(in: &state, at: pendingRelease)
                }
            } else {
                state.pendingRelease = time
                if configuration.debounceDuration == 0 {
                    closeInterval(in: &state, at: time)
                }
            }
        } else {
            state.pendingRelease = nil
        }

        states[stepID] = state
        currentLoadedDuration = loadedDuration(in: state, at: time)
    }

    mutating func pause(at workoutElapsed: TimeInterval) {
        guard let currentStepID, var state = states[currentStepID] else { return }
        closeInterval(in: &state, at: workoutElapsed)
        state.pendingStart = nil
        state.pendingRelease = nil
        state.permitsMergeWithPrevious = false
        states[currentStepID] = state
        currentLoadedDuration = loadedDuration(in: state, at: workoutElapsed)
    }

    mutating func endStep(
        stepID: String,
        at workoutElapsed: TimeInterval,
        status: WorkoutStepMeasurement.Status = .measured
    ) {
        guard var state = states[stepID] else { return }
        closeInterval(in: &state, at: workoutElapsed)
        state.pendingStart = nil
        state.pendingRelease = nil
        state.status = status
        states[stepID] = state
        currentStepID = stepID
        currentLoadedDuration = loadedDuration(in: state, at: workoutElapsed)
    }

    mutating func interrupt(at workoutElapsed: TimeInterval) {
        pause(at: workoutElapsed)
        guard let currentStepID, var state = states[currentStepID] else { return }
        state.status = .interrupted
        states[currentStepID] = state
    }

    mutating func finish(at workoutElapsed: TimeInterval) -> [WorkoutStepMeasurement] {
        for stepID in stepOrder {
            guard var state = states[stepID] else { continue }
            closeInterval(in: &state, at: workoutElapsed)
            state.pendingStart = nil
            state.pendingRelease = nil
            if state.sampleCount > 0, state.status == .unmeasured {
                state.status = .measured
            }
            states[stepID] = state
        }

        return stepOrder.compactMap { stepID in
            guard let state = states[stepID] else { return nil }
            return WorkoutStepMeasurement(
                stepID: stepID,
                plannedActiveDuration: state.plannedActiveDuration,
                intervals: state.intervals,
                peakLoadKGF: state.peakLoadKGF,
                sampleCount: state.sampleCount,
                status: state.status
            )
        }
    }

    private mutating func state(
        for stepID: String,
        plannedActiveDuration: TimeInterval,
        stepStartElapsed: TimeInterval
    ) -> StepState {
        if let state = states[stepID] { return state }

        let state = StepState(
            plannedActiveDuration: plannedActiveDuration,
            stepStartElapsed: stepStartElapsed
        )
        states[stepID] = state
        stepOrder.append(stepID)
        return state
    }

    private mutating func openInterval(in state: inout StepState, at time: TimeInterval) {
        let start = state.clippedTime(time)
        state.pendingStart = nil
        state.pendingRelease = nil

        if state.permitsMergeWithPrevious,
           let previous = state.intervals.last,
           start - previous.end <= configuration.mergeGapDuration {
            state.intervals.removeLast()
            state.openStart = previous.start
        } else {
            state.openStart = start
        }
        state.permitsMergeWithPrevious = true
    }

    private func closeInterval(in state: inout StepState, at time: TimeInterval) {
        guard let start = state.openStart else { return }
        let end = max(start, state.clippedTime(time))
        state.intervals.append(LoadInterval(start: start, end: end))
        state.openStart = nil
        state.pendingRelease = nil
    }

    private func loadedDuration(in state: StepState, at time: TimeInterval) -> TimeInterval {
        let closedDuration = state.intervals.reduce(0) { $0 + $1.duration }
        guard let openStart = state.openStart else { return closedDuration }
        return closedDuration + max(0, state.clippedTime(time) - openStart)
    }
}

File: Android/app/src/main/java/com/hangten/android/sensors/MotherboardWorkoutRecorder.kt
package com.hangten.android.sensors

data class MotherboardDetectionConfiguration(
    val thresholdKgf: Double = 2.5,
    val releaseRatio: Double = 0.8,
    val debounceMs: Long = 100,
    val mergeGapMs: Long = 150,
)

data class LoadInterval(val startMs: Long, val endMs: Long) {
    val durationMs: Long get() = (endMs - startMs).coerceAtLeast(0)
}

enum class MeasurementStatus { Measured, Unmeasured, Interrupted }

data class WorkoutStepMeasurement(
    val stepId: String,
    val plannedActiveDurationMs: Long,
    val intervals: List<LoadInterval>,
    val peakLoadKgf: Double?,
    val sampleCount: Int,
    val status: MeasurementStatus,
)

data class RecordedMeasurements(
    val measurements: List<MotherboardMeasurement>,
    val truncated: Boolean,
)

data class SensorWorkoutActivity(
    val profile: ForceSensorProfile,
    val measurements: List<MotherboardMeasurement>,
    val measurementsTruncated: Boolean,
    val steps: List<WorkoutStepMeasurement>,
)

class MeasuredWorkoutCollector {
    private val measurements = ArrayList<MotherboardMeasurement>(MAXIMUM_MEASUREMENT_COUNT)
    private var truncated = false

    fun append(measurement: MotherboardMeasurement) {
        if (measurements.size < MAXIMUM_MEASUREMENT_COUNT) measurements += measurement else truncated = true
    }

    fun complete(): RecordedMeasurements = RecordedMeasurements(measurements.toList(), truncated)

    companion object { const val MAXIMUM_MEASUREMENT_COUNT = 20_000 }
}

class MotherboardWorkoutRecorder(
    private val configuration: MotherboardDetectionConfiguration = MotherboardDetectionConfiguration(),
) {
    private data class State(
        val plannedDurationMs: Long,
        val stepStartMs: Long,
        val intervals: MutableList<LoadInterval> = mutableListOf(),
        var openStartMs: Long? = null,
        var pendingStartMs: Long? = null,
        var pendingReleaseMs: Long? = null,
        var peakLoadKgf: Double? = null,
        var sampleCount: Int = 0,
        var status: MeasurementStatus = MeasurementStatus.Unmeasured,
        var permitsMerge: Boolean = true,
    ) {
        fun clipped(timeMs: Long): Long = timeMs.coerceAtMost(stepStartMs + plannedDurationMs)
    }

    private val states = linkedMapOf<String, State>()
    private var currentStepId: String? = null

    fun consume(
        measurement: MotherboardMeasurement,
        stepId: String,
        plannedActiveDurationMs: Long,
        workoutElapsedMs: Long,
        stepStartElapsedMs: Long = workoutElapsedMs,
        isActive: Boolean,
    ) {
        currentStepId = stepId
        val state = state(stepId, plannedActiveDurationMs, stepStartElapsedMs)
        if (!isActive) return
        state.sampleCount += 1
        state.peakLoadKgf = maxOf(state.peakLoadKgf ?: measurement.aggregateLoadKgf, measurement.aggregateLoadKgf)
        val time = state.clipped(workoutElapsedMs)
        val qualifying = measurement.aggregateLoadKgf >= configuration.thresholdKgf
        val released = measurement.aggregateLoadKgf < configuration.thresholdKgf * configuration.releaseRatio
        if (state.openStartMs == null) {
            if (qualifying) {
                val pending = state.pendingStartMs
                if (pending != null && time - pending >= configuration.debounceMs) open(state, pending)
                else if (pending == null) {
                    state.pendingStartMs = time
                    if (configuration.debounceMs == 0L) open(state, time)
                }
            } else state.pendingStartMs = null
        } else if (released) {
            val pending = state.pendingReleaseMs
            if (pending != null && time - pending >= configuration.debounceMs) close(state, pending)
            else if (pending == null) {
                state.pendingReleaseMs = time
                if (configuration.debounceMs == 0L) close(state, time)
            }
        } else state.pendingReleaseMs = null
    }

    fun pause(atMs: Long) {
        currentStepId?.let(states::get)?.let { state ->
            close(state, atMs)
            state.pendingStartMs = null
            state.pendingReleaseMs = null
            state.permitsMerge = false
        }
    }

    fun endStep(stepId: String, atMs: Long, status: MeasurementStatus = MeasurementStatus.Measured) {
        states[stepId]?.let { state ->
            close(state, atMs)
            state.pendingStartMs = null
            state.pendingReleaseMs = null
            state.status = status
        }
    }

    fun interrupt(stepId: String, plannedActiveDurationMs: Long, stepStartElapsedMs: Long, atMs: Long) {
        val state = state(stepId, plannedActiveDurationMs, stepStartElapsedMs)
        close(state, atMs)
        state.pendingStartMs = null
        state.pendingReleaseMs = null
        state.permitsMerge = false
        state.status = MeasurementStatus.Interrupted
    }

    fun finish(atMs: Long): List<WorkoutStepMeasurement> = states.map { (stepId, state) ->
        close(state, atMs)
        state.pendingStartMs = null
        state.pendingReleaseMs = null
        if (state.sampleCount > 0 && state.status == MeasurementStatus.Unmeasured) state.status = MeasurementStatus.Measured
        WorkoutStepMeasurement(stepId, state.plannedDurationMs, state.intervals.toList(), state.peakLoadKgf, state.sampleCount, state.status)
    }

    private fun state(stepId: String, durationMs: Long, startMs: Long): State = states.getOrPut(stepId) {
        State(durationMs.coerceAtLeast(0), startMs)
    }

    private fun open(state: State, atMs: Long) {
        val start = state.clipped(atMs)
        state.pendingStartMs = null
        state.pendingReleaseMs = null
        val previous = state.intervals.lastOrNull()
        if (state.permitsMerge && previous != null && start - previous.endMs <= configuration.mergeGapMs) {
            state.intervals.removeAt(state.intervals.lastIndex)
            state.openStartMs = previous.startMs
        } else state.openStartMs = start
        state.permitsMerge = true
    }

    private fun close(state: State, atMs: Long) {
        val start = state.openStartMs ?: return
        state.intervals += LoadInterval(start, maxOf(start, state.clipped(atMs)))
        state.openStartMs = null
        state.pendingReleaseMs = null
    }
}

class SensorWorkoutRecorder(
    private val profile: ForceSensorProfile,
    configuration: MotherboardDetectionConfiguration = MotherboardDetectionConfiguration(),
) {
    private val samples = MeasuredWorkoutCollector()
    private val intervals = MotherboardWorkoutRecorder(configuration)

    fun consume(measurement: MotherboardMeasurement, stepId: String, plannedActiveDurationMs: Long, elapsedMs: Long, stepStartMs: Long, isActive: Boolean) {
        samples.append(measurement)
        intervals.consume(measurement, stepId, plannedActiveDurationMs, elapsedMs, stepStartMs, isActive)
    }

    fun pause(elapsedMs: Long) = intervals.pause(elapsedMs)

    fun complete(elapsedMs: Long): SensorWorkoutActivity {
        val recorded = samples.complete()
        return SensorWorkoutActivity(profile, recorded.measurements, recorded.truncated, intervals.finish(elapsedMs))
    }
}
